torch/testing/_internal/common_utils.py

```diff
@@ -1542,21 +1381,66 @@ def maybe_to_list(a: Any) -> Any:
             x = maybe_to_tensor(x)
             y = maybe_to_tensor(y)
 
-            if isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor):
-                self.assertEqual(
-                    x, y, atol=atol, rtol=rtol, msg=msg, exact_dtype=exact_dtype, exact_device=exact_device
-                )
+            if not (isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor)):
+                # In case we can't convert the inputs to a tensor, we fall back to comparing x and y as iterables
+                x = maybe_to_list(x)
+                y = maybe_to_list(y)
+
+            self.assertEqual(
+                x,
+                y,
+                msg=msg,
+                atol=atol,
+                rtol=rtol,
+                equal_nan=equal_nan,
+                exact_dtype=exact_dtype,
+                exact_device=exact_device,
+                exact_stride=exact_stride,
+                exact_is_coalesced=exact_is_coalesced,
+            )
+        elif isinstance(x, (Number, bool, np.bool_)) or isinstance(y, (Number, bool, np.bool_)):
+            # torch.testing.assert_close does not allow Tensor vs. scalar comparisons
+            dtype: Optional[torch.dtype]
+            if isinstance(x, torch.Tensor):
+                dtype = x.dtype
+            elif isinstance(y, torch.Tensor):
+                dtype = y.dtype
             else:
-                # In case we can't convert the array to a tensor, we fall back to comparing x and y as iterables
-                self.assertEqual(
-                    maybe_to_list(x),
-                    maybe_to_list(y),
-                    atol=atol,
-                    rtol=rtol,
-                    msg=msg,
-                    exact_dtype=exact_dtype,
-                    exact_device=exact_device
+                if isinstance(x, complex) or isinstance(y, complex):
+                    dtype = torch.complex128
+                elif isinstance(x, float) or isinstance(y, float):
+                    dtype = torch.float64
+                elif isinstance(x, int) or isinstance(y, int):
+                    dtype = torch.int64
+                elif isinstance(x, np.uint8) or isinstance(x, np.uint8):
+                    dtype = torch.uint8
+                elif isinstance(x, (bool, np.bool_)) or isinstance(y, (bool, np.bool_)):
+                    dtype = torch.bool
+                else:
+                    raise TypeError(f"Unknown types {type(x)} / {type(y)}")
+
+            try:
+                x = torch.as_tensor(x, dtype=dtype)
```

Comment:
won't this approach make 1.5 compare as equal to torch.tensor(1, dtype=torch.int32)?

Reply:
Yep, that is unfortunate.

Reply:
But can we make it so that doesn't happen?

Reply:
Yes, that needs to be handled. Since we are probably going to rework `assert_close` to handle Number vs Number, IMO we can change the behavior to convert the tensor to a Number rather then converting the Number to a Tensor. With that, the problem should go away.